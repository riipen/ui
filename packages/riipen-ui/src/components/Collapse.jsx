import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import { Transition } from "react-transition-group";
import css from "styled-jsx/css";

import ThemeContext from "../styles/ThemeContext";

const Collapse = props => {
  const {
    children,
    classes,
    component: Component = "div",
    in: inProp,
    ...other
  } = props;

  const theme = React.useContext(ThemeContext);

  const wrapperRef = React.useRef(null);

  const getLinkedStyles = () => {
    return css.resolve`
      .root {
        height: 0;
        min-height: 0px;
        overflow: hidden;
        transition: ${theme.transitions.create(["height"])};
      }

      .entered {
        height: auto;
        overflow: visible;
      }
      .exited {
        visibility: hidden;
      }
    `;
  };

  const getWrapperSize = () =>
    wrapperRef.current ? wrapperRef.current.clientHeight : 0;

  const handleEnter = node => {
    if (wrapperRef.current) {
      // Set absolute position to get the size of collapsed content
      wrapperRef.current.style.position = "absolute";
    }

    node.style.height = "0px";
  };

  const handleEntering = node => {
    const wrapperSize = getWrapperSize();

    if (wrapperRef.current) {
      // After the size is read reset the position back to default
      wrapperRef.current.style.position = "";
    }

    node.style.transitionDuration = `${theme.transitions.duration.standard}ms`;
    node.style.height = `${wrapperSize}px`;
  };

  const handleEntered = node => {
    node.style.height = "auto";
  };

  const handleExit = node => {
    node.style.height = `${getWrapperSize()}px`;
  };

  const handleExiting = node => {
    node.style.transitionDuration = `${theme.transitions.duration.standard}ms`;
    node.style.height = "0px";
  };

  const linkedStyles = getLinkedStyles();

  return (
    <Transition
      in={inProp}
      onEnter={handleEnter}
      onEntered={handleEntered}
      onEntering={handleEntering}
      onExit={handleExit}
      onExiting={handleExiting}
      timeout={theme.transitions.duration.standard}
      {...other}
    >
      {(state, childProps) => (
        <Component
          className={clsx(
            linkedStyles.className,
            "root",
            {
              entered: state === "entered",
              exited: state === "exited" && !inProp
            },
            classes
          )}
          {...childProps}
        >
          <div ref={wrapperRef}>{children}</div>
          {linkedStyles.styles}
        </Component>
      )}
    </Transition>
  );
};

Collapse.propTypes = {
  /**
   * The content node to be collapsed.
   */
  children: PropTypes.node,

  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.array,

  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component: PropTypes.elementType,

  /**
   * If `true`, the component will transition in.
   */
  in: PropTypes.bool
};

Collapse.defaultProps = {
  classes: [],
  component: "div"
};

Collapse.displayName = "Collapse";

export default Collapse;
