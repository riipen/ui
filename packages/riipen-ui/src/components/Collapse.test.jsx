import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Collapse from "./Collapse";

describe("<Collapse>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Collapse />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Collapse />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<Collapse />);

      expect(wrapper.find("Collapse").props().component).toEqual("div");
    });
  });

  describe("children prop", () => {
    it("renders given childen", () => {
      const child = <div>child</div>;

      const wrapper = mount(<Collapse>{child}</Collapse>);

      expect(wrapper.find("Collapse").contains(child)).toEqual(true);
    });
  });

  describe("classes prop", () => {
    it("applies given classes", () => {
      const classes = ["classOne"];

      const wrapper = mount(<Collapse classes={classes} />);

      expect(
        wrapper
          .find("Collapse")
          .childAt(0)
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Collapse classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("component prop", () => {
    it("sets component type as span", () => {
      const component = "span";

      const wrapper = mount(<Collapse component={component} />);

      expect(
        wrapper
          .find("Collapse")
          .childAt(0)
          .childAt(0)
          .type()
      ).toEqual(component);
    });
  });

  describe("in prop", () => {
    it("sets class name as entered when in prop is true", () => {
      const isTransition = true;

      const wrapper = mount(<Collapse in={isTransition} />);

      expect(
        wrapper
          .find("Collapse")
          .childAt(0)
          .childAt(0)
          .hasClass("entered")
      ).toEqual(true);
    });

    it("sets class name as exited when in prop is false", () => {
      const isTransition = false;

      const wrapper = mount(<Collapse in={isTransition} />);

      expect(
        wrapper
          .find("Collapse")
          .childAt(0)
          .childAt(0)
          .hasClass("exited")
      ).toEqual(true);
    });
  });
});
