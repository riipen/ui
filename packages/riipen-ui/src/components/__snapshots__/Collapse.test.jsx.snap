// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapse> renders correct snapshot 1`] = `
<Collapse
  classes={Array []}
  component="div"
>
  <Transition
    appear={false}
    enter={true}
    exit={true}
    in={false}
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={300}
    unmountOnExit={false}
  >
    <div
      className="jsx-2207988454 root exited"
    >
      <div />
      <JSXStyle
        dynamic={
          Array [
            "height 300ms cubic-bezier(0.4, 0, 0.2, 1)",
          ]
        }
        id="4123931595"
      />
    </div>
  </Transition>
</Collapse>
`;
